### PR TITLE
Dump out external trainable variable for external optimizer

### PR DESCRIPTION
**Pull request by tsarikahin**
closing #805 #395

### CODE REVIEW

**Comment by tsarikahin** [comment]
This merge request includes the following changes:

 - If an external optimizer function (e.g. L-BFGS) is used, it was not outputting or writing the external trainable variables after each epoch or in the training end. `In the training end` part was forgotten to implement.  Now one can output the external trainable variables after each epoch and at the very end as well
 - One example is tested if the algorithm works as expected

**Comment by lululxvi** [comment]
Format your code via [black](https://github.com/psf/black)

**Comment by tsarikahin** [comment]
@lululxvi I could not get why the tests could not pass? Is it because of formatting?

**Comment by lululxvi** [comment]
> @lululxvi I could not get why the tests could not pass? Is it because of formatting?

No. It seems an issue of TensorFlow installation in the Travis system.

File: deepxde/callbacks.py [hunk 1]
```diff
@@ -325,6 +325,9 @@ def on_epoch_end(self):
             self.epochs_since_last = 0
             self.on_train_begin()
 
+    def on_train_end(self):
```
Comment (lululxvi) [inline deepxde/callbacks.py:328]: Would adding this line makes two duplicated lines when training with Adam? Because the last result would be saved in both `on_epoch_end` and `on_train_end`.

File: deepxde/model.py [hunk 2]
```diff
@@ -617,6 +618,26 @@ def loss_callback(loss_train):
...
+                cb = self.callbacks.callbacks[0]
+                cb.epochs_since_last += 1
+
+                if cb.epochs_since_last >= cb.period:
+                    cb.epochs_since_last = 0
+
+                    print(
+                        cb.model.train_state.epoch,
+                        list_to_str(
+                            y[: len(self.external_trainable_variables)],
```
Comment (lululxvi) [inline deepxde/model.py:633]: There is no guarantee that the `external_trainable_variables` will be in the beginning.

File: deepxde/model.py [hunk 3]
```diff
@@ -629,6 +650,7 @@ def loss_callback(loss_train):
             feed_dict=feed_dict,
             fetches=[self.outputs_losses_train[1]],
             loss_callback=loss_callback,
+            step_callback=ext_var_callback,
```
Comment (lululxvi) [inline deepxde/model.py:653]: Do not use `step_callback`, as discussed above. Add the `external_trainable_variables` into `fetches`, and then use `loss_callback`.

File: examples/pinn_inverse/Lorenz_inverse.py [hunk 4]
```diff
@@ -73,9 +73,20 @@ def boundary(_, on_initial):
...
+
+# train adam
+model.compile(
+    "adam", lr=0.001, external_trainable_variables=external_trainable_variables
 )
-losshistory, train_state = model.train(iterations=60000, callbacks=[variable])
+losshistory, train_state = model.train(iterations=20000, callbacks=[variable])
+
+# train lbfgs
+model.compile("L-BFGS-B", external_trainable_variables=external_trainable_variables)
```
[the remark this answers is not in the document]
Comment (tsarikahin) [reply examples/pinn_inverse/Lorenz_inverse.py:89]: I am working on solid mechanic examples and here is the improvement:

image

As can be seen, there is a huge improvement with L-BFGS

File: deepxde/callbacks.py (hunk 1 above)
Comment (tsarikahin) [reply deepxde/callbacks.py:328]: This can be avoided by modifying the `on_epoch_end` function. If the epoch number is not the last epoch, then continue.

File: deepxde/model.py (hunk 2 above)
Comment (tsarikahin) [reply deepxde/model.py:633]: I thought they should be at the very end which might make sense. I run a couple of examples and it was always at the very beginning. How can we be sure? Is there any way to do that?

File: examples/pinn_inverse/Lorenz_inverse.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lululxvi) [reply examples/pinn_inverse/Lorenz_inverse.py:89]: Good. How about this Lorenz inverse problem you modified?

File: deepxde/callbacks.py (hunk 1 above)
Comment (lululxvi) [reply deepxde/callbacks.py:328]: Yes. You can modify it as you suggested.

File: deepxde/model.py (hunk 2 above)
Comment (lululxvi) [reply deepxde/model.py:633]: I don't know. That is why I suggest to use `loss_callback` as my next comment.

File: examples/pinn_inverse/Lorenz_inverse.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsarikahin) [reply examples/pinn_inverse/Lorenz_inverse.py:89]: Sorry, I was busy. Here are the results for Lorenz's problem. Initially, only ADAM is used and here are the results for c1, c2 and c3: (actually values 10, 15 and 8/3) 

```bash
19900 [9.74e+00, 1.50e+01, 2.67e+00]
19910 [9.74e+00, 1.50e+01, 2.67e+00]
19920 [9.74e+00, 1.50e+01, 2.67e+00]
19930 [9.74e+00, 1.50e+01, 2.67e+00]
19940 [9.74e+00, 1.50e+01, 2.67e+00]
19950 [9.74e+00, 1.50e+01, 2.67e+00]
19960 [9.74e+00, 1.50e+01, 2.67e+00]
19970 [9.75e+00, 1.50e+01, 2.67e+00]
19980 [9.75e+00, 1.50e+01, 2.67e+00]
19990 [9.75e+00, 1.50e+01, 2.67e+00]
20000 [9.75e+00, 1.50e+01, 2.67e+00]
20000 [9.75e+00, 1.50e+01, 2.67e+00]
```

Adam + LBFGS, it took 2608 steps more for LBFGS. And we can see that C1 is improved and it is perfectly matching with analytical results.

```bash
22426 [1.00e+01, 1.50e+01, 2.67e+00]
22436 [1.00e+01, 1.50e+01, 2.67e+00]
22447 [1.00e+01, 1.50e+01, 2.67e+00]
22457 [1.00e+01, 1.50e+01, 2.67e+00]
22467 [1.00e+01, 1.50e+01, 2.67e+00]
22477 [1.00e+01, 1.50e+01, 2.67e+00]
22487 [1.00e+01, 1.50e+01, 2.67e+00]
22497 [1.00e+01, 1.50e+01, 2.67e+00]
22507 [1.00e+01, 1.50e+01, 2.67e+00]
22517 [1.00e+01, 1.50e+01, 2.67e+00]
22527 [1.00e+01, 1.50e+01, 2.67e+00]
22537 [1.00e+01, 1.50e+01, 2.67e+00]
22547 [1.00e+01, 1.50e+01, 2.67e+00]
22558 [1.00e+01, 1.50e+01, 2.67e+00]
22568 [1.00e+01, 1.50e+01, 2.67e+00]
22578 [1.00e+01, 1.50e+01, 2.67e+00]
22588 [1.00e+01, 1.50e+01, 2.67e+00]
22608 [1.00e+01, 1.50e+01, 2.67e+00]
22608 [1.00e+01, 1.50e+01, 2.67e+00]
````

[the remark this answers is not in the document]
Comment (tsarikahin) [reply examples/pinn_inverse/Lorenz_inverse.py:89]: Also, the initial number of epochs is set to 60000, which is a huge number. I think adam + LBFGS has a huge improvement in training.

File: deepxde/callbacks.py (hunk 1 above)
Comment (tsarikahin) [reply deepxde/callbacks.py:328]: I can not reach the number of `epochs`, or iterations inside of the callbacks class. Therefore, I will introduce `num_iterations` in model class as None and assign that variable using epochs in training.

Comment (tsarikahin) [reply deepxde/callbacks.py:328]: Done.

File: deepxde/model.py (hunk 3 above)
Comment (tsarikahin) [reply deepxde/model.py]: Done.

File: examples/pinn_inverse/Lorenz_inverse.py (hunk 4 above)
Comment (lululxvi) [inline examples/pinn_inverse/Lorenz_inverse.py:89]: Replace "L-BFGS-B" with "L-BFGS"

File: deepxde/callbacks.py [hunk 5]
```diff
@@ -321,9 +322,14 @@ def on_train_begin(self):
 
     def on_epoch_end(self):
         self.epochs_since_last += 1
+        self.epochs_counter += 1
         if self.epochs_since_last >= self.period:
-            self.epochs_since_last = 0
-            self.on_train_begin()
+            if not (self.epochs_counter == self.model.num_iterations):
```
Comment (lululxvi) [inline deepxde/callbacks.py:327]: We don't need `self.epochs_counter` and `self.model.num_iterations` to check if it is the last step. We can use the following trick. In `on_train_end`, we check if `self.epochs_since_last == 0`: if it is true, then we don't run `self.on_train_begin()`, otherwise, we run it. You can double check the logic.

File: deepxde/model.py [hunk 6]
```diff
@@ -617,18 +619,39 @@ def loss_callback(loss_train):
                 )
                 display.training_display(self.train_state)
 
+            if self.external_trainable_variables:
+                cb = self.callbacks.callbacks[0]
```
Comment (lululxvi) [inline deepxde/model.py:623]: There is no guarantee `self.callbacks.callbacks[0]` is the varialbe callback. You should call `self.callbacks.on_epoch_end()`

Comment (tsarikahin) [reply deepxde/model.py]: Done.

File: deepxde/callbacks.py (hunk 5 above)
Comment (tsarikahin) [reply deepxde/callbacks.py]: Done.

File: examples/pinn_inverse/Lorenz_inverse.py (hunk 4 above)
Comment (tsarikahin) [reply examples/pinn_inverse/Lorenz_inverse.py]: Done.

**Comment by tsarikahin** [comment]
@lululxvi I could not comment on the file but line 562 in model.py I have comment.

```python
.
.
.
self.train_state.set_data_train(*self.data.train_next_batch(self.batch_size))
self.train_state.set_data_test(*self.data.test())
self._test()
self.callbacks.on_train_begin() # Line 562 
.
.
.
```
Line 562 also repeats the last epoch of the previous optimizer if more than one optimizer is used. I think it is not considered when it is implemented. I think this should be skipped if at least two optimizers are used. 

actually, line 562 is not needed at all. Because it is the initial guess before optimization starts. Do we really need it?

File: deepxde/model.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 from . import utils
 from .backend import backend_name, tf, torch, jax, paddle
 from .callbacks import CallbackList
+from .utils import list_to_str
```
Comment (lululxvi) [inline deepxde/model.py:17]: Delete

**Comment by lululxvi** [comment]
> Line 562 also repeats the last epoch of the previous optimizer if more than one optimizer is used. I think it is not considered when it is implemented and it should be skipped if at least two optimizers are used.
> 
> Actually, line 562 is not needed at all. Because it is the initial guess before optimization starts. Do we really need it?

Good point. I suggest keeping it, as the initial values are useful and we may tune the initial values in some cases. It is a future work to resolve this repeated output between two optimizers.

File: deepxde/model.py [hunk 8]
```diff
@@ -617,18 +617,24 @@ def loss_callback(loss_train):
...
+            fetches = [self.outputs_losses_train[1]] + self.external_trainable_variables
+        else:
+            fetches = [self.outputs_losses_train[1]]
         self.train_step.minimize(
             self.sess,
             feed_dict=feed_dict,
-            fetches=[self.outputs_losses_train[1]],
-            loss_callback=loss_callback,
+            fetches=fetches,
+            loss_callback=loss_callback
```
Comment (lululxvi) [inline deepxde/model.py:637]: `loss_callback=loss_callback,`

File: deepxde/model.py (hunk 7 above)
Comment (tsarikahin) [reply deepxde/model.py:17]: Done.

File: deepxde/model.py (hunk 8 above)
Comment (tsarikahin) [reply deepxde/model.py:637]: Done.

File: deepxde/model.py [hunk 9]
```diff
@@ -617,18 +617,24 @@ def loss_callback(loss_train):
                 )
                 display.training_display(self.train_state)
 
+            if self.external_trainable_variables:
+                self.callbacks.on_epoch_end()
```
[the remark this answers is not in the document]
Comment (tsarikahin) [reply deepxde/model.py:621]: Done.

File: deepxde/model.py [hunk 10]
```diff
@@ -617,18 +617,24 @@ def loss_callback(loss_train):
...
         self.train_state.set_data_train(*self.data.train_next_batch(self.batch_size))
         feed_dict = self.net.feed_dict(
             True,
             self.train_state.X_train,
             self.train_state.y_train,
             self.train_state.train_aux_vars,
         )
+        if self.external_trainable_variables:
+            fetches = [self.outputs_losses_train[1]] + self.external_trainable_variables
```
Comment (lululxvi) [inline deepxde/model.py:630]: Because in the callback, we recompute the variable values, do we need to fetch them here?

File: deepxde/model.py [hunk 11]
```diff
@@ -617,17 +616,23 @@ def loss_callback(loss_train):
                 )
                 display.training_display(self.train_state)
 
+                self.callbacks.on_batch_end()
+                self.callbacks.on_epoch_end()
```
Comment (lululxvi) [inline deepxde/model.py:620]: These two lines should be outside the `if self.train_state.step % display_every == 0:`. callback has its own period.

Comment (tsarikahin) [reply deepxde/model.py:620]: ahs sorry, yes.

File: deepxde/model.py (hunk 10 above)
Comment (tsarikahin) [reply deepxde/model.py:629]: If we don't fetch them, it will not update values so it will be always the last iteration of the adam optimizer.

File: deepxde/model.py [hunk 12]
```diff
@@ -620,22 +618,7 @@ def loss_callback(loss_train, *args):
...
-                        cb.model.train_state.epoch,
-                        list_to_str(
-                            [float(arg) for arg in args],
-                            precision=cb.precision,
-                        ),
-                        file=cb.file,
-                    )
-                    cb.file.flush()
-
+                self.callbacks.on_epoch_end()
```
Comment (tsarikahin) [inline deepxde/model.py:620]: It seems like this does not. Replacing lines 623-638 with `on_epoch_end` does not update the external variables so they don't change.

File: deepxde/model.py (hunk 10 above)
Comment (tsarikahin) [reply deepxde/model.py:629]: Can you please check the last comment?

File: deepxde/model.py (hunk 12 above)
Comment (lululxvi) [reply deepxde/model.py:620]: You are right. Then let us use the lines 623-638. You can check if there is a VariableValue callback.

Comment (tsarikahin) [reply deepxde/model.py]: Done.

**Comment by tsarikahin** [comment]
I have another remark regarding on outputting of the loss values in case of an external optimizer is used. Fetches contain only the training loss so it does not include the test loss. I think we should fetch the test losses as well right? It was lie this before, but I just realized it while plotting losses

File: deepxde/model.py [hunk 13]
```diff
@@ -616,18 +616,38 @@ def loss_callback(loss_train):
                     self.train_state.step, self.train_state.loss_train, None, None
                 )
                 display.training_display(self.train_state)
-
+            for cb in self.callbacks.callbacks:
+                if type(cb).__name__=="VariableValue":
```
Comment (lululxvi) [inline deepxde/model.py:620]: Refactor via black

File: deepxde/model.py [hunk 14]
```diff
@@ -616,18 +616,38 @@ def loss_callback(loss_train):
...
+                    cb = self.callbacks.callbacks[0]
```
Comment (lululxvi) [inline deepxde/model.py:621]: `cb = self.callbacks.callbacks[0]` why we need this?

**Comment by lululxvi** [comment]
> I have another remark regarding on outputting of the loss values in case of an external optimizer is used. Fetches contain only the training loss so it does not include the test loss. I think we should fetch the test losses as well right? It was like this before, but I just realized it while plotting losses

Yes, you are right.

File: deepxde/model.py (hunk 14 above)
Comment (tsarikahin) [reply deepxde/model.py]: Sorry, I missed it. Done.

**Comment by tsarikahin** [comment]
> > I have another remark regarding on outputting of the loss values in case of an external optimizer is used. Fetches contain only the training loss so it does not include the test loss. I think we should fetch the test losses as well right? It was like this before, but I just realized it while plotting losses
> 
> Yes, you are right.

Done.

File: deepxde/model.py [hunk 15]
```diff
@@ -624,10 +642,17 @@ def loss_callback(loss_train):
             self.train_state.y_train,
             self.train_state.train_aux_vars,
         )
+        if self.external_trainable_variables:
+            fetches = [
+                self.outputs_losses_train[1],
+                self.outputs_losses_test[1],
+            ] + self.external_trainable_variables
+        else:
+            fetches = [self.outputs_losses_train[1], self.outputs_losses_test[1]]
```
Comment (lululxvi) [inline deepxde/model.py:651]: test loss should also be added here.

Comment (lululxvi) [inline deepxde/model.py:651]:
```python
fetches = [self.outputs_losses_train[1], self.outputs_losses_test[1]]
if self.external_trainable_variables:
    fetches += self.external_trainable_variables
```
